Add a new app badge How-to page to PWA docs (#25143)
mdn/content: files/en-us/web/progressive_web_apps/how_to/display_badge_on_app_icon/index.md

---
title: Display a badge on the app icon
slug: Web/Progressive_web_apps/How_to/Display_badge_on_app_icon
---

> OS-native applications on mobile and desktop operating systems can display badges on top of their app icons to inform users that new content is available. For example, an email client application can display the total number of unread messages in a badge and update this number even if the app is not running.

Reviewer: ```suggestion
Applications native to the mobile and desktop operating systems can display badges on top of their app icons to inform users that new content is available. For example, an email client application can display the total number of unread messages in a badge and update this number even if the app is not running.
```
I didn't want to start article with an acronym - dashed word. Would be ok elsewhere.

[Progressive Web Apps](/en-US/docs/Web/Progressive_web_apps) (PWAs) can display and update badges on their app icons too.

> 

Reviewer: should we include screenshots? It might help newer readers know what we are talking about. But then it may also get dated super fast.

Author: Good point. I will add a screenshot at the very start to set the context a little better.

> Displaying a badge is done by using the [Badging API](/en-US/docs/Web/API/Badging_API), and updating the badge even when the app is not running can be done by using the app's [service worker](/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers).

Reviewer: ```suggestion
Displaying a badge is done by using the [Badging API](/en-US/docs/Web/API/Badging_API). Updating the badge even when the app is not running can be done by using the app's [service worker](/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers).
```

Reviewer 2: Does this imply that "displaying" and "updating" are two different things, and that only "updating" is a background SW activity?

Maybe:

```suggestion
Displaying and updating a badge is done by using the [Badging API](/en-US/docs/Web/API/Badging_API). You can call this API from the app's [service worker](/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers), to display or update the badge even when the app is not running.
```

?

Author: I like this, thanks both.

## Support for badges

App badges are only supported when a PWA is installed on its host operating system. Badges appear on the app icon which only exists after the app has been installed.

> > **Note:** that this article focuses on the {{domxref("Navigator.setAppBadge()")}} and {{domxref("Navigator.clearAppBadge()")}} methods from the Badging API and ignores the `Navigator.setClientBadge` and `Navigator.clearClientBadge`. Although these methods are defined in the [Badging API specification](https://w3c.github.io/badging/) too, they are not implemented in any browser and are meant to display badges on documents, not on application icons.

Reviewer: ```suggestion
> **Note:** This article focuses on the {{domxref("Navigator.setAppBadge()")}} and {{domxref("Navigator.clearAppBadge()")}} methods from the Badging API and ignores the `Navigator.setClientBadge` and `Navigator.clearClientBadge`. Although these methods are defined in the [Badging API specification](https://w3c.github.io/badging/) too, they are for displaying badges on documents, not on application icons.
```

> ### Desktop support

Reviewer: Should we create a BCD type table at the bottom or is there a BCD for desktop, browser, and PWA support? Either way, should we move the BCD-esque information to where BCD is normally found? I am sure we will hit this a lot as we write these pages, so thinking we should formalize.

Reviewer 2: There is BCD for `setAppBadge()` and `clearAppBadge()`:

https://developer.mozilla.org/en-US/docs/Web/API/Navigator/setAppBadge#browser_compatibility
https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clearAppBadge#browser_compatibility

This does include the support details that are described in this section of the new page.

I agree that we need a consistent way to represent support in these pages. In https://github.com/mdn/content/pull/24709 I already complained about using the `browser-compat` key, on the basis that this isn't a reference doc, but I wonder whether this isn't pointlessly pedantic. We could have a "Compatibility" or "Browser compatibility" H2, but instead of tables, have links to compat for any web platform features discussed, like:

> - [`navigator.setAppBadge`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/setAppBadge#browser_compatibility)
> - [`navigator.clearAppBadge`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clearAppBadge#browser_compatibility)

I think there might be some aspects of compat that are hard to capture in BCD, like [which browsers support installation at all](https://pr25058.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Guides/Making_PWAs_installable#browser_support) or which manifest members are mandatory for installation.

I'm going to ask for more opinions.

Author: Agreed that we need a consistent way to go about this. Having to write prose for which browsers support a given How-to is going to be hard to maintain too.
I originally added BCD tables in here (in another PR), but as discussed with Will, decided to remove them. I don't think having BCD tables, as they exist today, is the right thing to do for How-to pages. 
Say I want to write a How-to about how to download a movie file for offline viewing. This page will describe how to use Streams, IndexedDB, Service Workers, and more, all together. In this example, adding BCD tables would be kind of a mess because you'd see maybe 4 or 5 of them, each with the entire list of methods and events they contain, where in fact, you're only interested in a particular set of features.
We'd almost want to have a tool, based on BCD, that you can configure to spit out a custom BCD table that contains only what you need for a given page.

On desktop operating systems, badges are only supported on Windows and macOS, and only when the PWA is installed from Chrome or Edge. While the Badging API is supported on Chromium-based browsers on Linux, badges are not displayed on this operating system.

Safari and Firefox on desktop do not support the Badging API and do not support installing PWAs.

### Mobile support

On mobile operating systems, badges are supported on Chromium-based browsers running on Android.

## Badge best practices

Before learning how to use badges, consider these best practices to ensure your app uses badges in the most effective and useful way for your users.

### Check for support

> The Badging API is not [supported](#support-for-badges) in all browsers and all operating systems, so make sure you check for support in your JavaScript code before using the API:

Reviewer: ```suggestion
To ensure the Badging API is [supported](#support-for-badges) in the user's browsers and operating system, to prevent throwing a JavaScript error, check for support before using the API:
```

Reviewer: may need to change the link # based on the previous comment. 
Removing  browser/os support language to make the content last longer before needing to be revisited.

```javascript
if (navigator.setAppBadge) {
  // The API is supported, use it.
} else {
  // The API is not supported, don't use it.
}
```

> Additionally, even in browsers where it is supported, there are cases where using the API does not actually display a badge. For example, even if Chrome supports the Badging API, badges do not appear on installed application icons on Linux. Therefore, do not only rely on badges to inform users about the availability of new content.

Reviewer: ```suggestion
> **Note:** Do not rely solely on badges to inform users about the availability of new content. Browsers that support the Badging API may be installed on operating systems that do not support displaying a badge. For example, while Chrome supports the Badging API, badges will not appear on installed application icons on Linux. 
```

Author: I'm unsure about the performance of these notes and other similar warnings. In my experience working on docs, I feel like people have developed habits to skip over ads, and banners, and such, and focus only on the text content. 

I know these notes are made to attract attention, but I'm honestly unsure about whether they perform well.

Author: I'm definitely using your rewrite though, thanks!

### Use badges sparingly

> Like notifications, badges can be a very effective way to re-engage users with your app when used sparingly. Make sure to use badges to only signal new content that's important for your users to know about.

Reviewer: a link to an article about engagement would be fabulous here, if there is a good one.

### Update badges in real-time

> Make sure to update your application badge in real-time. This means updating the badge count to reflect how many new items are actually left for the user to consume, and clearing the app badge when there are no new items.

Reviewer: how is this done? If not code, a bullet list of pseudo code would help.

Author: I added some more prose. Let me know if that's ok.

> 
> ### Use badges sparingly
> 
> Like notifications, badges can be a very effective way to re-engage users with your app when used sparingly. Make sure to use badges to only signal new content that's important for your users to know about.
> 
> ### Update badges in real-time
> 
> Make sure to update your application badge in real-time. This means updating the badge count to reflect how many new items are actually left for the user to consume, and clearing the app badge when there are no new items.
> 
> ### Highlight new content in the app

Reviewer: ```suggestion

## Badging and user experience

While badging can increase engagement, make sure you only engage in ways that improve user experience. Consider the following recommendations:

- Use badges sparingly
  : - Like notifications, badges can be a very effective way to re-engage users with your app when used sparingly. Make sure to use badges to only signal new content that's important for your users to know about.
- Update badges in real-time
  : - Make sure to update your application badge in real-time. This means updating the badge count to reflect how many new items are actually left for the user to consume, and clearing the app badge when there are no new items.
- Highlight new content in the app
  - : When new content is available and signaled by the presence of a badge on the app icon, that new content should clearly be identifiable by the user when they launch the app.
```

> 
> When new content is available and signaled by the presence of a badge on the app icon, that new content should clearly be identifiable by the user when they launch the app.

Reviewer: ```suggestion
```
deleting because it's added in the previous suggestion

> For example, an email client app with an unread messages badge on the app icon should bold or highlight the unread messages when the app is opened.

Reviewer: do we want to touch on all three suggestions in the 'for example', such as "The badge itself should only appear if there is a message in the inbox, not for email filtered into the spam folder. "

Author: Thanks, I'm adding this as an example in the "Update badges in real time" section. This is a great suggestion.

## Displaying and updating the badge

To display a badge on your PWA's app icon that shows a number of unread messages, use the {{domxref("Navigator.setAppBadge()")}} method:

```javascript
// Check for support first.
if (navigator.setAppBadge) {
  // Display the number of unread messages.
  navigator.setAppBadge(numberOfUnreadMessages);
}

> ```

Reviewer: Should we add a note about updating as the number of unread messages goes down, not just at zero? That's not really 'how to', more adding hints to good UI practices.

Author: Thanks, I'm adding this as an example in the "Update badges in real time" section. This is a great suggestion.

PWAs can use the following mechanisms to update in the background and display, update, or hide their badges:

- [Push API](/en-US/docs/Web/API/Push_API)
  - : PWAs can use this API to receive messages from a server even when the app is not running.
- [Background Synchronization API](/en-US/docs/Web/API/Background_Synchronization_API)
  - : PWAs can use this API to run code in the background when a stable network connection is detected.
- [Web Periodic Background Synchronization API](/en-US/docs/Web/API/Web_Periodic_Background_Synchronization_API)
  - : PWAs can use this API to run code in the background at periodic intervals of time.

Here is a service worker code example showing how to listen to a server's Push messages and update the app badge to reflect an unread messages count:

```javascript
// Listen to "push" events in the service worker.
self.addEventListener("push", (event) => {
  // Extract the unread count from the push message data.
  const message = event.data.json();
  const unreadCount = message.unreadCount;

> 

Reviewer: do we need to add the support check?

Author: Oh yes, totally. Thanks for flagging this.
